`Modeling and Text Exercises`: View example solutions

terlan98 opened the pull request:
> ⚠️ Should be merged after #170 

This PR allows viewing example solutions for modeling and text exercises. The example solution is shown in CorrectionSidebar, below the problem statement.

terlan98 committed:
Enable text exercises

terlan98 committed:
Create TextSubmissionServiceImpl

terlan98 committed:
Make AssessmentView more generic

terlan98 committed:
Create a custom UndoManager to fix undo button bug

terlan98 committed:
Fix pencil mode and font size persisted among submissions

terlan98 committed:
Added mock exercise, submission, and AssessmentViewModel; Adjusted PaneVM for text exercises

terlan98 committed:
Move AssessmentVM and AssessmentResult ownership to AssessmentView; Move more logic into AssessmentVM

terlan98 committed:
Make CorrectionSidebarView and it's subviews independent of CodeEditorVM

terlan98 committed:
Refactor EditFeedback properties

terlan98 committed:
Display text exercise submission content

terlan98 committed:
Fix pencil mode toggle

terlan98 committed:
Load Participation (and feedbacks) for text exercises. Refactor some VM functions

terlan98 committed:
Show text highlights based on feedbacks; Remove unused cornerRadius from HighlightedRange; Fix dark mode issue

terlan98 committed:
Set ScorePicker maxScore value based on exercise; Fix one swiftlint warning

terlan98 committed:
Change highlight color based on inline feedback updates

terlan98 committed:
Implement feedback creation; Fix programming exercise feedback deletion issue;

terlan98 committed:
Implement feedback deletion for text exercises; Fix onDelete index error

terlan98 committed:
Add text selection with word-based granularity

terlan98 committed:
Limit word selection to at least 2 words; Prevent selecting words that are already highlighted

terlan98 committed:
Allow selecting a single word

terlan98 committed:
Change var to let

terlan98 committed:
Implement assessment saving

terlan98 committed:
Implement assessment cancellation

terlan98 committed:
Implement random assessment initialization

terlan98 committed:
Implement assessment submission and immediately starting the next assessment

terlan98 committed:
Fix incorrect max point label

terlan98 committed:
Fix incorrect score progress view

terlan98 committed:
Merge remote-tracking branch 'refs/remotes/origin/develop'

Conflicts:
	Themis.xcodeproj/project.pbxproj
	Themis/Views/Exams/ExamListView.swift
	Themis/Views/Exams/ExamSection.swift
	Themis/Views/Exams/ExamSectionDetailView.swift
	Themis/Views/Exams/ExamSectionView.swift
	Themis/Views/Exercises/ExerciseSection.swift
	Themis/Views/Exercises/ExerciseSectionsView.swift

terlan98 committed:
Remove local ArtemisCore dependency

terlan98 committed:
Merge branch 'develop' of https://github.com/ls1intum/Themis into feature/text-exercises

terlan98 committed:
Fix build error and feedback editing bug

terlan98 committed:
Fix undo problem with inline feedbacks

terlan98 committed:
Adjust link color for dark mode

terlan98 committed:
Fix formatting

terlan98 committed:
Change some variable types and access modifiers

terlan98 committed:
Remove unnecessary get function for BaseSubmission and its usages

terlan98 committed:
Add more documentation; Fix a preview

terlan98 committed:
Change more vars into lets

terlan98 committed:
Improve error handling

terlan98 committed:
Change HighlightedRange id type to UUID

terlan98 committed:
Improve optional course id and exercise id handling for problem statement

terlan98 committed:
Handle the case when no submissions are left to assess

terlan98 committed:
Remove commented code

terlan98 committed:
Enable modeling exercises; Implement rendering logic (partially) for class diagrams

terlan98 committed:
Detect tapped UML elements

terlan98 committed:
Highlight model elements on tap

terlan98 committed:
Add a new canvas to render highlights

terlan98 committed:
Implement package drawing

terlan98 committed:
Align text within uml elements

terlan98 committed:
Implement rendering logic for all relationship types

terlan98 committed:
Add UML background tile

terlan98 committed:
Add type annotation to class diagram elements

terlan98 committed:
Remove unnecessary lines

terlan98 committed:
Remove unnecessary lines

terlan98 committed:
Merge branch 'feature/modeling-exercises' of https://github.com/ls1intum/Themis into feature/modeling-exercises

terlan98 committed:
Merge branch 'feature/modeling-exercises' of https://github.com/ls1intum/Themis into feature/modeling-exercises

terlan98 committed:
Simplify switch statement

terlan98 committed:
Implement submission fetching

terlan98 committed:
Merge remote-tracking branch 'refs/remotes/origin/feature/modeling-exercises'

terlan98 committed:
Implement general feedback actions and Save action.

terlan98 committed:
Implement assessment cancellation

terlan98 committed:
Implement random new submission fetching

terlan98 committed:
Extract common fields from TextExerciseRendererViewModel and CodeEditorViewModel into a parent view model class

terlan98 committed:
Show referenced feedback badges on UML elements

terlan98 committed:
Show referenced feedback badges on UML relationships

terlan98 committed:
Merge branch 'develop' of https://github.com/ls1intum/Themis into feature/modeling-exercises

terlan98 committed:
Implement highlight editing and deletion

terlan98 committed:
Implement referenced feedback creation

terlan98 committed:
Make relationships selectable for referenced feedback

terlan98 committed:
Implement undo/redo actions

terlan98 committed:
Change selected element highlight style

terlan98 committed:
Implement assessment submission; Fix pencil mode issues; Make some UI-related functions run on MainActor

terlan98 committed:
Add more UML arrow directions; Fix left-facing arrow position;

terlan98 committed:
Draw multiplicities

terlan98 committed:
Draw class diagram roles

terlan98 committed:
Fix class diagram relationship highlight appearance

terlan98 committed:
Fix class diagram relationship tap gesture

terlan98 committed:
Fix class diagram relationship tap gesture

terlan98 committed:
Merge branch 'feature/modeling-exercises' of https://github.com/ls1intum/Themis into feature/modeling-exercises

terlan98 committed:
Improve badge placement for relationships

terlan98 committed:
Resolve some warnings

terlan98 committed:
Fix SwiftLint issues

terlan98 committed:
Update ArtemisCore dependency

terlan98 committed:
Implement getAllSubmissions in ModelingSubmissionServiceImpl

terlan98 committed:
Implement read-only submissions; Generalize AssessmentViewModel and create subclasses

terlan98 committed:
Fix a bug related to starting a new assessment after submitting one

terlan98 committed:
Draw suggested feedback highlights

terlan98 committed:
Add hint box for automatic feedback

terlan98 committed:
Remove filtering of automatic feedbacks with zero points

terlan98 committed:
Fix wrong arrowhead

terlan98 committed:
Merge branch 'develop' of https://github.com/ls1intum/Themis into feature/modeling-exercises

terlan98 committed:
Add temporary highlight that is shown when a feedback cell is tapped

terlan98 committed:
Deselect UML item when edit sheet is closed

terlan98 committed:
Remove hint box background

terlan98 committed:
Reduce warnings

terlan98 committed:
Better error handling for unsupported diagram types

terlan98 committed:
Add corresponding classess for use case diagram; Restructure UML-related folders;

terlan98 committed:
Draw relationships

terlan98 committed:
Draw relationship type text

terlan98 committed:
Draw relationship names

terlan98 committed:
Rotate relationship type text to keep it aligned with the arrow

terlan98 committed:
Refactoring

terlan98 committed:
Improve relationship higlights

terlan98 committed:
Implement drag gesture

terlan98 committed:
Add padding to Canvas

terlan98 committed:
Implement magnification gesture

terlan98 committed:
Add reset button for zoom and position

terlan98 committed:
Suppress SwiftLint type_body_length in UMLRendererViewModel

terlan98 committed:
Add example solution to modeling and text assessment

terlan98 committed:
Add center button to example model

terlan98 committed:
Improve error handling

terlan98 committed:
Make diagram text react to font size changes

terlan98 committed:
Remove local ArtemisCore dependency; Fix element selection bug

terlan98 committed:
Add stroke to badge icons and make their background darker

terlan98 committed:
Change badge background color

terlan98 committed:
Add padding to problem statement

terlan98 committed:
rename athenaHintBox to automaticFeedbackHintBox

terlan98 committed:
Remove unnecessary tab in AssessmentViewModel

terlan98 committed:
Implement method and attribute grouping

terlan98 committed:
Reduce swiftlint warnings

terlan98 committed:
Reduce more SwiftLint warnings

terlan98 committed:
Remove problem statement padding for programming exercises

terlan98 committed:
Merge remote-tracking branch 'refs/remotes/origin/feature/modeling-exercises'

terlan98 committed:
Merge remote-tracking branch 'refs/remotes/origin/feature/modeling/use-case-diagrams'

Conflicts:
	Themis/ViewModels/Assessment/Modeling/UMLRendererViewModel.swift

terlan98 committed:
Fix UML diagram size issue after merge

terlan98 committed:
Merge branch 'feature/modeling/gestures' of https://github.com/ls1intum/Themis into feature/example-solutions

terlan98 committed:
Merge branch 'develop' of https://github.com/ls1intum/Themis into feature/modeling/use-case-diagrams

terlan98 committed:
Merge branch 'develop' into feature/modeling/use-case-diagrams

maximiliansoelch committed:
Merge branch 'develop' into feature/modeling/use-case-diagrams

maximiliansoelch committed:
Merge branch 'develop' into feature/modeling/use-case-diagrams

terlan98 committed:
Merge branch 'feature/modeling/use-case-diagrams' into feature/modeling/gestures

maximiliansoelch committed:
Merge branch 'develop' into feature/modeling/use-case-diagrams

terlan98 committed:
Merge branch 'feature/modeling/gestures' into feature/example-solutions

terlan98 committed:
Reduce SwiftLint warnings

terlan98 committed:
Merge branch 'feature/modeling/use-case-diagrams' into feature/modeling/gestures

terlan98 committed:
Apply suggestions from code review

Co-authored-by: Maximilian Sölch <maximiliansoelch@users.noreply.github.com>

terlan98 committed:
Replace annotation titles of UMLElementType

terlan98 committed:
Fix SwiftLint warning

terlan98 committed:
Merge branch 'bugfix/courseview-grouping' of https://github.com/ls1intum/Themis into feature/text/new-athena

terlan98 committed:
Merge branch 'develop' into feature/modeling/gestures

terlan98 committed:
Merge branch 'feature/modeling/gestures' into feature/example-solutions

terlan98 committed:
Remove UMLUseCaseDiagramRenderer and UMLClassDiagramRenderer initializers

terlan98 committed:
Add canvasBounds and fontSize to UMLDiagramRenderer protocol

terlan98 committed:
Restore deleted TextAssessmentView

terlan98 committed:
Fix diagram movement while dragging the right pane

terlan98 committed:
Merge branch 'feature/modeling/gestures' into feature/example-solutions

terlan98 committed:
Apply suggestions from code review

Co-authored-by: Nityananda Zbil <nityananda.zbil@icloud.com>

terlan98 committed:
Merge branch 'feature/modeling/use-case-diagrams' into feature/modeling/gestures

terlan98 committed:
Merge remote-tracking branch 'refs/remotes/origin/feature/modeling/gestures'

Conflicts:
	Themis/Views/Assessment/Modeling Exercise/UMLRenderer.swift

terlan98 committed:
Fix warning & merge conflict

terlan98 committed:
Merge remote-tracking branch 'refs/remotes/origin/develop'

Conflicts:
	Themis/Views/Assessment/Modeling Exercise/UMLRenderer.swift

terlan98 committed:
Merge branch 'develop' of https://github.com/ls1intum/Themis into feature/example-solutions

terlan98 committed:
Merge branch 'develop' into feature/example-solutions

maximiliansoelch commented on Themis/ViewModels/Assessment/Modeling/UMLRendererViewModel.swift:
```diff
@@ -113,6 +137,18 @@ class UMLRendererViewModel: ExerciseRendererViewModel {
+        if let point {
+            currentDragLocation = point
+        } else {
+            currentDragLocation = .init(x: diagramSize.height - 50, // 50 – default padding added by Athena
```
What does the padding for modeling exercises have to do with Athena?

terlan98 replied:
Nothing, I wanted to write Apollon 😅

maximiliansoelch committed:
Merge branch 'develop' into feature/example-solutions

maximiliansoelch commented on Themis/Views/Assessment/CorrectionSidebar/CorrectionSidebarView.swift:
```diff
@@ -35,16 +35,19 @@ struct CorrectionSidebarView: View {
+                    .opacity(correctionSidebarStatus == .problemStatement ? 1.0 : 0.0001)
+                    // 0.0 causes this view to be redrawn and webview to send a new request
+                    
+                    if !assessmentVM.loading {
```
Is the second loading check required? 🤔

terlan98 replied:
Actually, the first check is unnecessary. I removed that and kept this one.

maximiliansoelch reviewed (changes requested):
Is this intended?

https://github.com/ls1intum/Themis/assets/9954674/ae8f0e13-c6c9-4c09-ba00-0701bce42650

Also, the example solution does not support skeleton loading.

terlan98 committed:
Fix wrong comment about Athena

terlan98 committed:
Show skeleton for example solutions; Remove redundant loading state check

terlan98 commented:
Thanks for the review, @maximiliansoelch 👍 
I addressed your comments. 

Regarding the video: I think it's not an issue because if the fingers are located on different canvases during the **pinch** gesture, it becomes two separate **drag** gestures, affecting both canvases. Here's a video showing this behavior in more detail:

https://github.com/ls1intum/Themis/assets/22798314/945b5706-4d3b-4f76-ac18-d767408cd035

terlan98 committed:
Merge remote-tracking branch 'refs/remotes/origin/feature/example-solutions'

terlan98 committed:
Small refactoring for the SkeletonExtension
